plugins/thunderbolt/fu-thunderbolt-image-validation.c

```diff
@@ -0,0 +1,482 @@
+} FwLocation;
+
+typedef struct {
+	const guint8 *data;
+	gsize len;
+	guint32 *sections;
+} FwObject;
+
+typedef struct {
+	guint16 id;
+	guint gen;
+	guint ports;
+} HwInfo;
+
+static HwInfo *
+get_hw_info (guint16 id)
+{
+	static const HwInfo hw_info_arr[] = {
+		{ 0x156D, 2, 2 }, /* FR 4C */
+		{ 0x156B, 2, 1 }, /* FR 2C */
+		{ 0x157E, 2, 1 }, /* WR */
+
+		{ 0x1578, 3, 2 }, /* AR 4C */
+		{ 0x1576, 3, 1 }, /* AR 2C */
+		{ 0x15C0, 3, 1 }, /* AR LP */
+		{ 0x15D3, 3, 2 }, /* AR-C 4C */
+		{ 0x15DA, 3, 1 }, /* AR-C 2C */
+
+		{ 0 }
+	};
+
+	for (int i = 0; hw_info_arr[i].id != 0; i++)
+		if (hw_info_arr[i].id == id)
+			return hw_info_arr + i;
+	return NULL;
+}
+
+static inline gboolean
+valid_farb_pointer (guint32 pointer)
+{
+	return pointer != 0 && pointer != 0xFFFFFF;
+}
+
+static GByteArray *
+read_location (const FwLocation *location, const FwObject *fw)
+{
+	guint32 location_start = fw->sections[location->section] + location->offset;
+	g_autoptr(GByteArray) read = g_byte_array_new ();
+
+	// TODO: set error?
+	if (location_start > fw->len || location_start + location->len > fw->len)
+		return NULL;
+
+	read = g_byte_array_append (read,
+				    fw->data + location_start,
+				    location->len);
+
+	if (location->mask)
+		read->data[0] &= location->mask;
+	return g_steal_pointer (&read);
+}
+
+static guint32
+read_farb_pointer_impl (const FwLocation *location, const FwObject *fw)
+{
+	g_autoptr(GByteArray) farb = read_location (location, fw);
+	farb = g_byte_array_set_size (farb, sizeof(guint32));
+	memset (farb->data + location->len, 0, farb->len - location->len);
+	return GUINT32_FROM_LE (*(guint32*) farb->data);
+}
+
+static guint32
+read_farb_pointer (const FwObject *fw)
+{
+	const FwLocation farb1 = { .offset = 0,      .len = 3 };
+	const FwLocation farb2 = { .offset = 0x1000, .len = 3 };
+
+	guint32 ret = read_farb_pointer_impl (&farb1, fw);
+	if (valid_farb_pointer (ret))
+		return ret;
+	return read_farb_pointer_impl (&farb2, fw);
+}
+
+static gboolean
+compare (const FwLocation *location, const FwObject *controller_fw, const FwObject *image_fw)
+{
+	g_autoptr(GByteArray) controller_data = read_location (location, controller_fw);
+	g_autoptr(GByteArray) image_data      = read_location (location, image_fw);
+
+	if (!image_data || !controller_data ||
+	    controller_data->len < location->len ||
+	    image_data->len < location->len)
+		return FALSE;
+
+	return memcmp (controller_data->data, image_data->data, location->len) == 0;
+}
+
+static gboolean
+read_bool (const FwLocation *location, const FwObject *fw)
+{
+	g_autoptr(GByteArray) read = read_location (location, fw);
+	for (gsize i = 0; i < read->len; i++)
+		if (read->data[i] != 0)
+			return TRUE;
+	return FALSE;
+}
+
+static guint16
+read_uint16 (const FwLocation *location, const FwObject *fw)
+{
+	g_autoptr(GByteArray) read = read_location (location, fw);
+	g_assert_cmpuint (location->len, ==, sizeof (guint16));
+	return GUINT16_FROM_LE (*(guint16*) read->data);
```

Comment:
This looks exciting from an ARM/PPC point of view. I think you might have to memcpy the location into a temporary guint26 to avoid aliased pointer warnings.

Reply:
I have this at mind when I wrote this, meant to change it (or at least ask about it) and forgot :(
Sure, alignment issues are undefined behavior. Fixing.